Serialising

toschneider opened the pull request.

toschneider committed:
wip serializing network

toschneider committed:
implements serialiazation using serde_json

toschneider committed:
Merge branch 'master' into serialising

toschneider committed:
adding comments

toschneider commented:
closes #7

vab9 commented on data/config.json:
```diff
@@ -0,0 +1 @@
+{"layers":[4,20,3],"weights":[[20,4,[-0.38057694,-0.021769967,-1.2681402,0.27779853,0.42083523,-1.5866951,0.029875442,0.42735842,0.82491314,-1.1873777,0.5607671,-0.47871846,-0.97258115,0.39857262,0.46550456,0.6525046,1.2224816,-0.54037476,0.7059634,-1.3471562,0.17896108,-0.27190933,-1.6669575,-0.26766515,1.4962062,-2.4499378,0.8107155,0.2532993,-0.14283562,-1.7229908,-0.29564518,-0.8089977,1.6552 ...
```
Ich glaube der Inhalt dieses data/ directory sollten wir nicht ins repo pushen...

vab9 replied:
Das gleiche gilt fuer data/config.xml...
Hast du `git add .` gemacht?

vab9 commented on src/main.rs:
```diff
@@ -1,13 +1,15 @@
 extern crate rand;
+extern crate nalgebra;
```
Wir hatten uns auf `extern crate nalgebra as na;` geeinigt...

vab9 commented on src/main.rs:
```diff
@@ -53,7 +55,20 @@ fn main() {
           nn.get_biases().len());
 
 
+    println!("FF: {:?}",
```
Bitte nur noch den Logger zur Ausgabe benutzen. Sonst weinen Johann und ich.

vab9 commented on src/main.rs:
```diff
@@ -53,7 +55,20 @@ fn main() {
-          nn.feedforward(na::DVector::from_element(nn.get_layers()[0] as usize, 0 as f32)));
-    info!("ended");
+          nn.feedforward(nalgebra::DVector::from_element(nn.get_layers()[0] as usize, 0 as f32)));
+        info!("ended");
```
Hier stimmt irgendwas mit der Einrückung nicht.

vab9 commented on src/main.rs:
```diff
@@ -53,7 +55,20 @@ fn main() {
+          nn.feedforward(nalgebra::DVector::from_element(nn.get_layers()[0] as usize, 0 as f32)));
+        info!("ended");
+
+        nn.serialize(&path.join(&config));
```
Hier auch, oder? Kann das bei github nicht so gut erkennen...

vab9 commented on src/main.rs:
```diff
@@ -53,7 +55,20 @@ fn main() {
+        nn.serialize(&path.join(&config));
+    let tt = match nn::Network::deserialize(&path.join(&config)) {
+        Ok(val) => val,
+        Err(e) => { println!("{:?}",e);
```
Logger, kein println bitte

vab9 commented on src/nn/mod.rs:
```diff
@@ -1,7 +1,14 @@
 use rand::Rng;
 use rand::distributions::normal::StandardNormal;
-use na::{DMatrix, DVector, IterableMut};
+use nalgebra::{DMatrix, DVector, IterableMut};
```
`use na::..`

vab9 commented on src/nn/mod.rs:
```diff
@@ -91,6 +98,92 @@ impl Network {
+    pub fn serialize(&self, filename: &Path) {
+        let mut b = Vec::new();
+        let biases = self.biases.clone();
+        // parses `Vec<DVector<f32>>` into `Vec<Vec<f32>>`
```
@toschneider Wenn wir das eh von einem DVector in einen Vec überspielen, dann brauchen wir doch nicht mehr serialize/deserialize implementieren und können stattdessen einfach den Vec serialisieren, oder?

toschneider replied:
genau in dem Fall reicht dann derive

vab9 commented on src/nn/mod.rs:
```diff
@@ -91,6 +98,92 @@ impl Network {
+        let t: Net = match serde_json::from_str(&data) {
+            Ok(val) => val,
+            Err(e) => {
+                println!("{:?}",e);
```
Logger

vab9 commented on src/nn/mod.rs:
```diff
@@ -91,6 +98,92 @@ impl Network {
+        let j = match serde_json::to_string(&t) {
+            Ok(val) => val,
+            Err(e) => {
+                println!("{:?}",e);
```
logger

vab9 commented on src/nn/mod.rs:
```diff
@@ -91,6 +98,92 @@ impl Network {
+    pub fn deserialize(filename: &Path) -> Result<Network, &'static str> {
+        // reads state from file into string
+        let mut data = String::new();
+        let f = File::open(filename).expect("Unable to open file");
```
Vielleicht besser mit ? als mit expect.

vab9 commented on src/nn/mod.rs:
```diff
@@ -91,6 +98,92 @@ impl Network {
+        // writes data into given file
+        let f = File::create(filename).expect("Unable to create file");
+        let mut f = BufWriter::new(f);
+        f.write_all(j.as_bytes()).expect("Unable to write data");
```
Return type von deserialize ist doch ein Result, dann benutz doch `?` anstatt expect

vab9 commented on src/nn/mod.rs:
```diff
@@ -91,6 +98,92 @@ impl Network {
+        let mut data = String::new();
+        let f = File::open(filename).expect("Unable to open file");
+        let mut br = BufReader::new(f);
+        br.read_to_string(&mut data).expect("Unable to read string");
```
`?`

vab9 commented on src/nn/mod.rs:
```diff
@@ -91,6 +98,92 @@ impl Network {
+        br.read_to_string(&mut data).expect("Unable to read string");
+
+        // parses string into network container
+        let t: Net = match serde_json::from_str(&data) {
```
Kann glaube ich auch mit `?` geloest werden.

vab9 commented on src/structs/flower.rs:
```diff
@@ -1,7 +1,6 @@
 use structs::Classifier;
 use std::fmt;
-use na::DVector;
+use nalgebra::DVector;
```
`use na::DVector;`

vab9 commented on src/structs/mod.rs:
```diff
@@ -1,7 +1,7 @@
 
 use structs::flower::{Flower, FlowerName};
-use na::DVector;
+use nalgebra::DVector;
```
`use na::DVector;`

vab9 commented on src/nn/mod.rs:
```diff
@@ -91,6 +98,92 @@ impl Network {
+
+        let mut w = Vec::new();
+        let weights = self.weights.clone();
+        // parses `Vec<DMatrix<f32>>` into `Vec<(usize, usize, Vec<f32>)>`
```
Haste dir mal https://docs.rs/nalgebra/0.10.1/nalgebra/struct.DMatrix.html#method.into_vector angeguckt? Sollte damit deutlich einfacher gehen.

vab9 reviewed (changes requested):
Habe einiges angemerkt. Die Umwandlung in ein neuen Container ist zwar blöd, aber wenn es anders nicht funktioniert dann ist das ok. Allerdings brauchen wir in dem Fall nicht mehr Serialize/Deserialize für unser Struct zu implementieren. Um DMatrizen/DVectoren in klassische Vec<> umzuwandeln gibt es in nalgebra hilfsmethoden bzw. ist ein one-liner mit Iteratoren (siehe weiter unten). Dann brauchen wir im Code einfach nur dieses Wrapper Struct mit `#[derive()]` zu serialisieren...

vab9 committed:
add ability to choose log level

gustav1101 committed:
Merge pull request #35 from vab9/log-verbosity

Seems ok. Merging log verbosity.

joarndt committed:
solved merge conflict and add config to .gitignore

joarndt committed:
added config.json to gitignore

vab9 committed:
try to ignore config.json

vab9 committed:
WIP serializing things

vab9 committed:
finally resolve #7 - allow saving a nn to a file

With the new save_to_file() and from_file() functions it is now very easy to serialize and deserialize a nn state.
To that end, we introduce a Serializable wrapper to the Network structure.
- Also added a util mod that for now only has one method, which gets the project root directory
- minor tweaks and cleanups
